pkg/proctree/proctree_procfs.go

```diff
@@ -15,6 +15,32 @@ const (
 	AllPIDs = 0
 )
 
+func (pt *ProcessTree) feedFromProcFSLoop() {
+	pt.procfsChan = make(chan int, 1000)
+	go func() {
+		for {
+			select {
+			case <-pt.ctx.Done():
+				return
+			case givenPid := <-pt.procfsChan:
+				err := pt.FeedFromProcFS(givenPid)
+				if err != nil {
+					logger.Debugw("proctree from procfs", "err", err)
+				}
+			}
+		}
+	}()
+}
+
+// FeedFromProcFSAsync feeds the process tree with data from procfs asynchronously.
+func (pt *ProcessTree) FeedFromProcFSAsync(givenPid int) {
+	if pt.procfsChan == nil {
+		logger.Debugw("starting procfs proctree loop") // will tell if called more than once
+		pt.feedFromProcFSLoop()
+	}
+	pt.procfsChan <- givenPid // feed the loop
```

Comment:
This line might prove blocking in a very busy environment.
I propose to do the following:
```Go
select {
case pt.procfsChan <- givenPid:  // feed the loop
    break
default:
    logger.Warnw("proctree loop feeding channel is full")
}
return
```

Reply:
Though maybe if we create a warning it should be written once or once per some period

Reply:
```go
func (pt *ProcessTree) feedFromProcFSLoop() {
	go func() {
		for {
			select {
			case <-time.After(15 * time.Second):
				pt.procfsOnce = new(sync.Once) // reset the once every 15 seconds
			case <-pt.ctx.Done():
				return
			case givenPid := <-pt.procfsChan:
				err := pt.FeedFromProcFS(givenPid)
				if err != nil {
					logger.Debugw("proctree from procfs", "err", err)
				}
			}
		}
	}()
}

// FeedFromProcFSAsync feeds the process tree with data from procfs asynchronously.
func (pt *ProcessTree) FeedFromProcFSAsync(givenPid int) {
	if pt.procfsChan == nil {
		logger.Debugw("starting procfs proctree loop") // will tell if called more than once
		pt.procfsChan = make(chan int, 1000)
		pt.feedFromProcFSLoop()
	}
	if pt.procfsOnce == nil {
		pt.procfsOnce = new(sync.Once)
	}

	// feed the loop without blocking (if the loop is busy, given pid won't be processed)
	select {
	case pt.procfsChan <- givenPid: // feed the loop
		break
	default:
		pt.procfsOnce.Do(func() {
			// only log once if the loop is busy (avoid spamming the logs), once is reset every 15s
			logger.Debugw("procfs proctree loop is busy")
		})
	}
}
```

Reply:
Nice!